@grafana/ui: Create slider component

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
Creates reusable and configurable slider component that is going to be needed for Rich History. 
![image](https://user-images.githubusercontent.com/30407135/74756007-3c535680-5274-11ea-960c-bbca2092ccc5.png).

PR adds test coverage + adds component to Storybook. 

![ezgif com-video-to-gif (3)](https://user-images.githubusercontent.com/30407135/74781509-e39ab280-52a1-11ea-8097-6ceab4c90d34.gif)

**Special notes for your reviewer**:

At the end, I have decided to use **rc-slider** (and not **react-slider** as initially thought), as **rc-slider** is going to give us more versatility, if we are going to need it in the future, compared to **react-slider**.

ivanahuckova committed:
grafana/ui: Create slider

ivanahuckova committed:
grafana/ui: Create slider, tests, add to storybook

torkelo commented on packages/grafana-ui/src/components/Slider/Slider.test.tsx:
```diff
@@ -0,0 +1,18 @@
+describe('Slider', () => {
+  it('renders correct html', () => {
+    const wrapper = mount(<Slider {...sliderProps} />);
+    expect(wrapper.html()).toMatchSnapshot();
```
We have stopped using snapshots as they are mostly just annoying, better to test that the render did not fail. And check for something important you expect in the rendered output

torkelo commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,65 @@
+  reverse?: boolean;
+  size?: string;
+  formatTooltipResult?: (value: any) => number | string;
+  onChange?: (value: any) => void;
```
since min & max numbers should not value be a number?

torkelo commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,65 @@
+  max: number;
+  orientation: 'horizontal' | 'vertical';
+  reverse?: boolean;
+  size?: string;
```
what does this do? what are the possible values?

ivanahuckova replied:
I will add the comment to make it more clear and change it to number to be even more clear. But basically, it sets width of horizontal slider and height of vertical slider. So you can change the size based on your usecase. 🙂

ivanahuckova replied:
The default value is 200px.

ivanahuckova replied:
I have changes the naming to lengthOfSlider, so it is more clear. Also, I have added possibility to customize lengthOfSlider in storybook.
![image](https://user-images.githubusercontent.com/30407135/74764792-ad9a0600-5282-11ea-8664-45746bee4895.png)

torkelo commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,65 @@
+  );
+};
+
+export const Slider = withTheme(UnThemedSlider);
```
For functional components do not use withTheme, you can use the useTheme() hook instead inside the function component, example: https://github.com/grafana/grafana/blob/eba0390502f39276f657f21e4270b2b2008c1df0/packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx#L15

torkelo reviewed (commented):
Great job! will be super useful in the new panel edit mode!

ivanahuckova committed:
Update Slider, minor changes

torkelo commented:
<img width="793" alt="Screen Shot 2020-02-18 at 20 20 25" src="https://user-images.githubusercontent.com/10999/74770362-5ef16980-528c-11ea-9f67-924dd7531c9f.png">

The track color is too dark, almost invisible in dark & light theme. 

Currently, you can define two values. We will need a way to only define a single value as well so we can use it for options like line width,  point size etc.

ivanahuckova committed:
Implement single value slider

ivanahuckova commented:
> The track color is too dark, almost invisible in dark & light theme.
Updated:
![image](https://user-images.githubusercontent.com/30407135/74782855-9835d380-52a4-11ea-8d50-f84f4b47295e.png)
![image](https://user-images.githubusercontent.com/30407135/74782918-c0bdcd80-52a4-11ea-8388-55579e1190c3.png)

> Currently, you can define two values. We will need a way to only define a single value as well so we can use it for options like line width, point size etc.

Added option for single value:
![image](https://user-images.githubusercontent.com/30407135/74783012-fa8ed400-52a4-11ea-86c1-a2ec2a56113b.png)

ivanahuckova committed:
Update style

dprokop commented on packages/grafana-ui/package.json:
```diff
@@ -76,6 +77,7 @@
     "@types/node": "10.14.1",
     "@types/papaparse": "4.5.9",
     "@types/pretty-format": "20.0.1",
+    "@types/rc-slider": "^8.6.5",
```
```suggestion
    "@types/rc-slider": "8.6.5",
```

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.story.tsx:
```diff
@@ -0,0 +1,39 @@
+import { Slider } from './Slider';
+import { select, number, boolean } from '@storybook/addon-knobs';
+
+const getStory = (title: string, component: any) => ({
```
Not sure why you need this?  I think introducing such function should happen in `utils/storybook` and should allow some more customisation of the stories, like `parameters` customisation.

In this case default exporting an object should be enough.

ivanahuckova replied:
Yes, you are right, we don't need it atm. I  was thinking, that if we add storybook slider docs in the future, it would be ready for it, but now it is redundant. So I've removed it.

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,69 @@
+  min: number;
+  max: number;
+  orientation: 'horizontal' | 'vertical';
+  // Set current positions of handle(s). If only 1 value supplied, only 1 handle displayed.
```
```suggestion
  /** Set current positions of handle(s). If only 1 value supplied, only 1 handle displayed. */
```

dprokop replied:
@ivanahuckova having `//` comments will result in docs not showing in props table. You should use `/** ... */` comments in order for the description to show in storybook docs:
![image](https://user-images.githubusercontent.com/2376619/74909760-4468de80-53b9-11ea-9c7b-bdf72fffce1f.png)


I'm working on styleguide update that covers this: https://github.com/grafana/grafana/pull/22197/files#diff-d7a6487cc6d2e9460188e8527b28839eR204

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,69 @@
+  // Set current positions of handle(s). If only 1 value supplied, only 1 handle displayed.
+  value?: number[];
+  reverse?: boolean;
+  lengthOfSlider?: number;
```
I'm not a huge fan of statically defined length. Can we instead make the slider be always 100% and adjust the width with the container element when necessary?

ivanahuckova replied:
Yeah, you are right. I was thinking about it, but then I've decided to add defined length. I have changed it to 100%.

dprokop replied:
thank you 🙏

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,69 @@
+export interface Props {
+  min: number;
+  max: number;
+  orientation: 'horizontal' | 'vertical';
```
Can we make this optional and default to `horizontal`?

ivanahuckova replied:
Yes, great point. Updated!

dprokop commented on packages/grafana-ui/src/components/Slider/_Slider.scss:
```diff
@@ -0,0 +1,57 @@
+@import '../../node_modules/rc-slider/assets/index.css';
+
+$blue: #1f60c4;
```
isn't this $blue-shade from the theme?

dprokop commented:
Looks really good! One minor UI thing I have noticed is that tooltip can be used to drag the range. But the cursor on it defaults to text selection. Maybe would be good to change it hand.

ivanahuckova committed:
Update packages/grafana-ui/package.json

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Update packages/grafana-ui/src/components/Slider/Slider.tsx

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Update slider, include PR review feedback

ivanahuckova commented:
> I have noticed is that tooltip can be used to drag the range. But the cursor on it defaults to text selection. Maybe would be good to change it hand.

@dprokop You have the best  👀 for detail! Great catch, updated.

ivanahuckova committed:
Merge branch 'ivanah/grafana-ui-slider' of https://github.com/grafana/grafana into ivanah/grafana-ui-slider

dprokop commented:
One minor thing left UI wise. Text of the tooltip should not be selectable other wise it can result in this:
![image](https://user-images.githubusercontent.com/2376619/74908719-ee933700-53b6-11ea-8e99-8905eb671541.png)

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,71 @@
+import { GrafanaTheme } from '@grafana/data';
+import { useTheme } from '../../themes/ThemeContext';
+
+export enum SliderOrientation {
```
I've created a very similar type yesterday (https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/components/Layout/Layout.tsx#L6) but I'm not happy with it. Can you please unify these two types to a single one and place in grafana-ui/types?

Also, I have second thoughts about using `enum`. I know it's tempting, but it has some consequences later on when consuming the API. I.e. you cannot use string literals when applying the component:

```tsx
<Slider orientation="horizontal" ... />
```

will not work (technically it will, but TS will complain), and in order to make it work you will need to import the `SliderOrientation` enum from grafana-ui too:

```tsx
import { Slider, SliderOrientation } from '@grafana/ui';

<Slider orientation={SliderOrientation.horizontal} />
```

Instead just define type:  
```ts
type Orientation = 'horizontal' | 'vertical';
```

---
I'd rather avoid the need to import enums from @grafana/ui and I'd like above(type instead of enum) to become a pattern we use. Some thougts on this @torkelo ?

torkelo replied:
Yea, enums can be good but for props like this I find a string type like:
type Orientation = 'horizontal' | 'vertical';

better as you can use the component without importing more types.

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,71 @@
+        margin: ${theme.spacing.sm} ${theme.spacing.lg} ${theme.spacing.sm} ${theme.spacing.sm};
+      `;
+  return {
+    container: container,
```
```suggestion
    container,
```

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,71 @@
+  onChange?: (values: number[]) => void;
+}
+
+const defaultProps = { orientation: SliderOrientation.horizontal };
```
You don't need to define default props for components like this. Instead, you can assign default values when destructing the properties, like this:

```
export const Slider: FunctionComponent<Props> = ({
  min,
  max,
  onChange,
  orientation = 'horizontal',
  reverse,
  formatTooltipResult,
  value,
})
```

dprokop reviewed (approved):
This is very close to be mergable. Left some final nit comments :)

ivanahuckova committed:
Update packages/grafana-ui/src/components/Slider/Slider.tsx

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Update packages/grafana-ui/src/components/Slider/Slider.tsx

Co-Authored-By: Dominik Prokop <dominik.prokop@grafana.com>

ivanahuckova committed:
Export orientatin types, gix selectability of tooltip text

peterholmberg commented on packages/grafana-ui/src/components/Slider/_Slider.scss:
```diff
@@ -0,0 +1,60 @@
+@import '../../node_modules/rc-slider/assets/index.css';
+
+.rc-slider-tooltip {
```
It would be nice to have these overrides in emotion so we can be consistant with other components. See TimePickerCalendar for an example.

ivanahuckova replied:
That is a good comment! Honestly, I am not sure which one is preferred, because I've seen examples of both and I felt that this would be more preferred way, as we still have to create _Slider.scss to import css from _node_modules/rc-slider/assets/index.css_, so we just can use it to store overrides as well. But if it is not, I can definitely move overrides to emotion.

ivanahuckova replied:
Also, when I looked at the components, that have css file, the overrides are usually part of that file.

peterholmberg replied:
Yeah, we talked about this in our daily and we concluded that we can keep the import in the scss file but move the overrides to the  component (in emotion)

torkelo replied:
Also talked about adding a css loader to our webpack config, then we can import the css file from code as well so no need to have a scss file

peterholmberg reviewed (changes requested):
Nice work, just one comment about styles.

ivanahuckova committed:
Remove Orientation export from grafana/ui

ivanahuckova committed:
Testing Global component to inject global styles

ivanahuckova committed:
Merge remote-tracking branch 'origin/master' into ivanah/grafana-ui-slider

dprokop commented on packages/grafana-ui/src/components/Slider/Slider.tsx:
```diff
@@ -0,0 +1,124 @@
+        border: 1px solid ${trackColor};
+      }
+    `,
+    tooltip: cssCore`
```
@ivanahuckova please leave some comment about this usage in the code, ok? Same about `Global` below :)

ivanahuckova committed:
Add comments

peterholmberg reviewed (approved):
LGTM!
